### ISSUE TITLE
ec2_group failure When proto is either 50 or 51

### ISSUE THREAD

**Issue by rainmanh** [ansible/ansible#11177]
Hi

I am having the following issue when applying the variable set below:

```
ec2_security_groups_test:
    - sg_name: test-grup
      sg_description: test-group
      vpc_id: "vpc-XXXXXX"
      region: "{{ region }}"
      sg_rules:
       - proto: 50
         from_port: -1
         to_port: -1
         cidr_ip: 12.12.12.12/32
       - proto: 50
         from_port: -1
         to_port: -1
         cidr_ip: 13.13.13.13/32
       - proto: 50
         from_port: -1
         to_port: -1
         group_name: test-group

```

it will work on the 1st run, but on the second I will have:

```
<Response><Errors><Error><Code>InvalidPermission.Duplicate</Code><Message>the specified rule "peer: 12.12.12.12/32, protocol: 50, ALLOW" already exists</Message></Error></Errors><RequestID>e890595b-9609-4c31-a611-87da7b5de7ae</RequestID></Response>


```

I have tried specifying Ports, also "-1".. Nothing seems to work.
If I say TCP or UDP it works fine, but when I specify a numeric port as above I got the issue:

Any possible workaround?

**Comment by bennojoy** [comment]
@rainmanh  from teh boto docs the boto protocol only accepts strings for protocol so it shoudl be one of "tcp/udp/icmp"   and from port and to ports takes the range. so -1 is not a valid value afaik, so if you want to allow all port for tcp , the protocol : tcp from_port: 0 to_port: 65536 woudl be the way to go.

**Comment by bennojoy** [comment]
for icmp -1 is valid though, and seems to work if call multiple times also

**Comment by rainmanh** [comment]
Then this is an issue purely from boot?
It is strange I can run an insert of 50 and 51 but it fails after when rerunning the same insert. Boto seems accepting the rule omn the 1st round.
Any way of maybe skipping those proto rules (50/51) if already found ?

The error is basically saying it is trying to re-create the rule when it already exists, so I do no think that is down to boot but a bit more parsing can be integrated:

```
  File "ansible-tmp-1433479180.89-23107385894344/ec2_group", line 304, in main
    group.authorize(rule['proto'], rule['from_port'], rule['to_port'], ip, grantGroup)
  File "/usr/local/lib/python2.7/dist-packages/boto/ec2/securitygroup.py", line 203, in authorize
    dry_run=dry_run)

```

**Comment by bennojoy** [comment]
@rainmanh yeah you are right, seems like we do allow protocols other than tcp/udp/icmp, and it is failing as it thinks that rule doesnt exist. will check.

**Comment by bennojoy** [comment]
@rainmanh  can you please give #1472  a try.

**Comment by rainmanh** [comment]
Hi @bennojoy 
I have applied 
`````
lib/ansible/runner/__init__.py
result_data = self._executor(hosts[0]).result
`````
I cannot find
``````
action_plugins/pause.py
hosts = ', '.join(self.runner.host_set)
```````

Still I run it to see what happens but the same result...

``````
invalid output was: Traceback (most recent call last):
  File "/var/lib/jenkins/.ansible/tmp/ansible-tmp-1433512587.35-262629230190204/ec2_group", line 1936, in <module>
    main()
  File "/var/lib/jenkins/.ansible/tmp/ansible-tmp-1433512587.35-262629230190204/ec2_group", line 309, in main
    group.authorize(rule['proto'], rule['from_port'], rule['to_port'], ip, grantGroup)
  File "/usr/local/lib/python2.7/dist-packages/boto/ec2/securitygroup.py", line 203, in authorize
    dry_run=dry_run)
  File "/usr/local/lib/python2.7/dist-packages/boto/ec2/connection.py", line 3192, in authorize_security_group
    params, verb='POST')
  File "/usr/local/lib/python2.7/dist-packages/boto/connection.py", line 1223, in get_status
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidPermission.Duplicate</Code><Message>the specified rule "peer: 12.12.12.12/32, protocol: 50, ALLOW" already exists</Message></Error></Errors><RequestID>dcb45acf-3ab6-4acd-9d82-b20af3649565</RequestID></Response>


FATAL: all hosts have already failed -- aborting
```````

**Comment by bennojoy** [comment]
@rainmanh  how did you apply the patch, we shouldn be getting runner errors.

can you create a directory "library" in the same level as your playbook and place this file there and check.

https://raw.githubusercontent.com/bennojoy/ansible-modules-core/fix_11177/cloud/amazon/ec2_group.py

**Comment by rainmanh** [comment]
Hey @bennojoy 

My mistake. I have removed the previous post as it was completely irrelevant! For some reason I applied something completely different

I have applied https://github.com/ansible/ansible-modules-core/pull/1472 
to /usr/share/ansible/cloud/ec2_group 
(This is an ubuntu box)
And it seems to work perfectly fine.

I am writing some extra rules now so if anything I will let you know

I will get back to my post in google so also everybody knows about it

Thanks!
